scalafix-rules/src/main/scala/scalafix/internal/rule/ExplicitResultTypes.scala

```diff
@@ -92,35 +92,32 @@ final class ExplicitResultTypes(
     }
   }
 
-  override def fix(implicit ctx: SemanticDocument): Patch = {
+  override def fix(implicit ctx: SemanticDocument): Patch =
     try unsafeFix()
     catch {
-      case _: CompilerException =>
-        shutdownCompiler()
-        global.restart()
-        try unsafeFix()
-        catch {
-          case _: CompilerException if !config.fatalWarnings =>
-            // Ignore compiler crashes unless `fatalWarnings = true`.
-            Patch.empty
-        }
+      case _: CompilerException if !config.fatalWarnings =>
+        Patch.empty
     }
-  }
+
   def unsafeFix()(implicit ctx: SemanticDocument): Patch = {
-    lazy val types = TypePrinter(global.value, config)
-    ctx.tree.collect {
-      case t @ Defn.Val(mods, Pat.Var(name) :: Nil, None, body)
-          if isRuleCandidate(t, name, mods, body) =>
-        fixDefinition(t, body, types)
-
-      case t @ Defn.Var(mods, Pat.Var(name) :: Nil, None, Some(body))
-          if isRuleCandidate(t, name, mods, body) =>
-        fixDefinition(t, body, types)
-
-      case t @ Defn.Def(mods, name, _, _, None, body)
-          if isRuleCandidate(t, name, mods, body) =>
-        fixDefinition(t, body, types)
-    }.asPatch
+    global.value match {
+      case Some(value) =>
+        val types = new CompilerTypePrinter(value, config)
+        ctx.tree.collect {
+          case t @ Defn.Val(mods, Pat.Var(name) :: Nil, None, body)
+              if isRuleCandidate(t, name, mods, body) =>
+            fixDefinition(t, body, types)
+
+          case t @ Defn.Var(mods, Pat.Var(name) :: Nil, None, Some(body))
+              if isRuleCandidate(t, name, mods, body) =>
+            fixDefinition(t, body, types)
+
+          case t @ Defn.Def(mods, name, _, _, None, body)
+              if isRuleCandidate(t, name, mods, body) =>
+            fixDefinition(t, body, types)
+        }.asPatch
+      case None => Patch.empty
+    }
```

Comment:
old code vs new code:
 - in old code, we will always traverse the entire Tree of the file even if we were not able to instantiate the global, which can only happen if the classpath is wrong
 - new code: we always instantiate the global, even if there no element (def, or var or val) in the Tree. But in the other hand we remove an indirection where we create  a `TypePrinter ` just to wrap the case where ScalafixGlobal is None.

Reply:
I am not familiar with `global`, but https://github.com/scalacenter/scalafix/commit/c6014cebbbbea44b196d53991822c9418a469a6f did introduce this for something else than an incorrect classpath it seems